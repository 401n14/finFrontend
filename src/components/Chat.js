import React, { useState, useEffect } from "react";
import "../styles/chat.scss";
import useSockets from "./useSockets";
const fetch = require('node-fetch');

const apiKey = process.env.GOOGLE_API_KEY;

let options = {
  concurrentLimit: 20,
  requestOptions: {},
};

const googleTranslate = require('google-translate')(apiKey, options);


function Chat() {
  const [name, setName] = useState(null);
  const [welcome, setWelcome] = useState("Welcome!");
  const [message, setMessage] = useState(null);
  const [language, setLanguage] = useState("en");
  const [groupMessage, setGroupMessage] = useState(null);
  const { socket, socketVal, isConnected } = useSockets(
    "https://jamesdunn-lab23.herokuapp.com/",
    "broadcast"
  );


  const getTranslation = async () => {
      let res = await fetch('http://localhost:3000/detect?language=' + language);
      let json = await res.text();
      console.log('JSON RESULTS: ', json);
  }

  const sendGreeting = () => {
    if (name) setWelcome(`Hi ${name}! Welcome to the Chat!`);
  };

  const handleEnter = e => {
    if (e.key === "Enter") {
      e.target.className === "name-input" ? setUserName(e) : sendMessage(e);
    }
  };

  const selectLanguage = async e => {
    await googleTranslate.detectLanguage(e.target.value, async function(err, detection) {
        // if unable to detect a language >> default to english
        if (!detection) {
          setLanguage('en');
        } else {
          setLanguage(detection.language);
        }
  });
}

  const sendMessage = e => {
    e.preventDefault();
    if (message) {
      socket.emit("message", { name, message });
    }
    setMessage('');
  };

  const setUserName = e => {
    e.preventDefault();
    sendGreeting();
    getTranslation();
  };

  useEffect(() => {
      if (socketVal.name && socketVal.message) {
        googleTranslate.translate(socketVal.message, language, function (err, translation) {
            /**
                       * sends the translated message consisting of '{user: user, color: data.color, message: translation.translatedText}'
                       * @event message
                       * @param {string} message message event
                       * @param {object} translationData {user: object, color: string, message, string}
                       * @param {object} user user object
                       * @param {string} data.color color from the data object
                       * @param {string} translation.translatedText translated version of the text
                       */
                      setGroupMessage(`${socketVal.name}:  ${translation.translatedText}`);
            });
        }
  }, [socketVal, language]);

  return (
    <div className="Chat">
      <h1 className="chat-heading">{welcome}</h1>
      <h3>
        {isConnected
          ? "You are connected to the chat"
          : "You are not connected to the chat"}
      </h3>
      <div className="name-section">
        <button className="name-button" onClick={setUserName}>
          Submit Name
        </button>
        <input
          type="text"
          className="name-input"
          onKeyUp={handleEnter}
          onChange={e => {
            setName(e.target.value);
          }}
        />
      </div>
      <div>
        <select onChange={selectLanguage}>
          <option>Select Language</option>
          <option>English</option>
          <option>Espanol</option>
        </select>
      </div>
      <div className="chat-section">
        <button className="chat-button" onClick={sendMessage}>
          Send Message
        </button>
        <input
          type="text"
          className="chat-input"
          value={message}
          onKeyUp={handleEnter}
          onChange={e => {
            setMessage(e.target.value);
          }}
        />
      </div>
      <p className="chat-output">{groupMessage}</p>
    </div>
  );
}

export default Chat;
