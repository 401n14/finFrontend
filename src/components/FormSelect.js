import React from 'react';

function FormSelect(props) {
    let options = [];

    options = props.list.map((item, ind)=> {
      return (
        <option key={ind} value={item.name}>
          {item.name}
        </option>
      );
    });
  
    return (
      <select
        onChange={props.onChange}
      >
          <option> -- Select Language -- </option>
        {options}
      </select>
    );
}

export default FormSelect;