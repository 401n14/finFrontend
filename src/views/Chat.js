import React, { useState, useEffect, useCallback } from "react";
import useSockets from "../utils/useSockets";
import { useAuth0 } from "../react-auth0-spa.js";
import FormSelect from "../components/FormSelect";
import data from '../components/data/data';

const fetch = require('node-fetch');

function Chat() {
  const [name, setName] = useState(null);
  const [welcome, setWelcome] = useState("Welcome!");
  const [message, setMessage] = useState(null);
  const [language, setLanguage] = useState("English");
  const [translation, setTranslation] = useState("en");
  //----------------New= []      Old=null -------------- Updated to old to fix Error
  const [groupMessage, setGroupMessage] = useState(null);
  const { socket, socketVal, isConnected } = useSockets(
    "https://jamesdunn-lab23.herokuapp.com/",
    "broadcast"
  );
  const { user } = useAuth0();

  const getLanguage = useCallback (
      async () => {
      let res = await fetch('https://translation-server.herokuapp.com/detect?language=' + language);
      let json = await res.text();
      setTranslation(json);
  }, [language]);

  const translateMessage = useCallback(
      async data => {
      let res = await fetch('https://translation-server.herokuapp.com/translate?message=' + data.message + '&translation=' + data.translation);
      let json = await res.text();
      socketVal.message = await json;
      //------------------New Code Caused Constant message firing--------------------
  //     let msg = [...groupMessage];
  //     msg.push(<p>`${socketVal.name}:  ${socketVal.message}`</p>)
      
  //     setGroupMessage(msg);
  
  // }, [socketVal.message, socketVal.name, groupMessage]);

  //--------------------- Old code I reverted to----------
  setGroupMessage(`${socketVal.name}:  ${socketVal.message}`);	
}, [socketVal.message, socketVal.name]);
//-----------------------End Old code
  const handleEnter = e => {
    if (e.key === "Enter") {
      sendMessage(e);
    }
  };

  const sendMessage = e => {
    e.preventDefault();
    if (message) {
      socket.emit("message", { name, message });
    }
    setMessage('');
  };

  useEffect(() => {
    setName(user.nickname);
    setWelcome(`Hi ${name}! Welcome to the Chat!`)
    
  }, [name, user.nickname]);

  useEffect(() => {
    getLanguage();
  }, [getLanguage, language])

  useEffect(() => {
      if (socketVal.name && socketVal.message) {
          (async () => {
        await translateMessage({message: socketVal.message, translation});
          })();
        
        }
  }, [socketVal, translateMessage, translation]);

  return (
    <div className="Chat">
      <h1 className="chat-heading">{welcome}</h1>
      <h3>
        {isConnected
          ? "You are connected to the chat"
          : "You are not connected to the chat"}
      </h3>

      <FormSelect list={data.Languages}  onChange={e => {
            setLanguage(e.target.value)}} />

      <div className="chat-section">
        <button className="chat-button" onClick={sendMessage}>
          Send Message
        </button>
        <input
          type="text"
          className="chat-input"
          value={message}
          onKeyUp={handleEnter}
          onChange={e => {
            setMessage(e.target.value);
          }}
        />
      </div>
      <p className="chat-output">{groupMessage}</p>
    </div>
  );
}

export default Chat;