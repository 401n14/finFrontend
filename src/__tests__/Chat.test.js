import React from 'react';
import Chat from '../views/Chat';
import {configure, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { useAuth0 } from '../react-auth0-spa.js';
import FormSelect from "../components/FormSelect";
import Data from '../components/data/data';
import renderer from 'react-test-renderer';

configure({adapter:new Adapter()});

const user = {
    email: 'johndoe@me.com',
    email_verified: true,
    sub: 'google-oauth2|2147627834623744883746',
  };

  jest.mock('../react-auth0-spa.js');

describe('Chat Tests', () => {
    beforeEach(() => {
      // Mock the Auth0 hook and make it return a logged in state
      useAuth0.mockReturnValue({
        isAuthenticated: true,
        user
      });
    });
    it('Renders with required props', async () => {
      const wrapper = mount(<Chat />);
      expect(wrapper).toBeTruthy();
    });
    it('renders correctly', () => {
        const page = renderer.create(<Chat />).toJSON();
        expect(page).toMatchSnapshot();
    });
    it('changes state message on change', async () => {
      const wrapper = mount(<Chat />);
      let message = wrapper.find('input.chat-input');
      message.simulate('change', {target: {value: 'hello'}});

      expect(message.instance().value).toEqual('hello');
    });
    it('renders language selector', async () => {
      const wrapper = mount(<Chat
      />)
      let language = wrapper.find('select').first();
      language.simulate('change', {target: {value: 'English'}});
      expect(language.instance().value).toEqual('-- Select Language --');
    });
  });