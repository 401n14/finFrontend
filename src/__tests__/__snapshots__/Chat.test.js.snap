// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Tests renders correctly 1`] = `
<div
  className="Chat"
>
  <h1
    className="chat-heading"
  >
    Welcome!
  </h1>
  <h3>
    You are not connected to the chat
  </h3>
  <select
    onChange={[Function]}
  >
    <option>
       -- Select Language -- 
    </option>
    <option
      value="English"
    >
      English
    </option>
    <option
      value="Espanol"
    >
      Espanol
    </option>
    <option
      value="日本語"
    >
      日本語
    </option>
    <option
      value="한국어"
    >
      한국어
    </option>
  </select>
  <div
    className="chat-section"
  >
    <button
      className="chat-button"
      onClick={[Function]}
    >
      Send Message
    </button>
    <input
      className="chat-input"
      onChange={[Function]}
      onKeyUp={[Function]}
      type="text"
      value=""
    />
  </div>
  <p
    className="chat-output"
  />
</div>
`;
